storage/innobase/dict/dict0dd.cc

```diff
@@ -6082,4 +6083,111 @@ bool dd_is_table_in_encrypted_tablespace(const dict_table_t *table) {
+@param[in] update       function object that will be invoked for updating DD
+flags
+@return false on success */
+static bool dd_update_tablespace_dd_flags(
+    THD *thd, const char *space_name, std::function<void(uint32 &)> update) {
+  Disable_autocommit_guard autocommit_guard(thd);
+  dd::cache::Dictionary_client *client = dd::get_dd_client(thd);
+  dd::cache::Dictionary_client::Auto_releaser releaser(client);
+
+  dd::Tablespace *dd_space = nullptr;
+
+  trx_t *trx = check_trx_exists(thd);
+  trx_start_if_not_started(trx, true);
+
+  if (dd::acquire_exclusive_tablespace_mdl(thd, space_name, false)) {
+    ut_a(false);
+  }
+
+  if (client->acquire_for_modification<dd::Tablespace>(space_name, &dd_space) ||
+      dd_space == nullptr) {
+    return (true);
+  }
+
+  uint32_t dd_space_flags = 0;
+
+  if (dd_space->se_private_data().get(dd_space_key_strings[DD_SPACE_FLAGS],
+                                      &dd_space_flags)) {
+    return (true);
+  }
+
+  update(dd_space_flags);
+
+  /* Update DD flags for tablespace */
+  dd_space->se_private_data().set(dd_space_key_strings[DD_SPACE_FLAGS],
+                                  static_cast<uint32>(dd_space_flags));
+
+  if (dd::commit_or_rollback_tablespace_change(thd, dd_space, false)) {
+    return (true);
+  }
+
+  return (false);
+}
+
+bool dd_set_encryption_flag(THD *thd, const char *space_name) {
+  auto update_func = [](uint32_t &dd_space_flags) {
+    dd_space_flags |= (1U << FSP_FLAGS_POS_ENCRYPTION);
+  };
+  return dd_update_tablespace_dd_flags(thd, space_name, update_func);
+}
+
+bool dd_clear_encryption_flag(THD *thd, const char *space_name) {
+  auto update_func = [](uint32_t &dd_space_flags) {
+    dd_space_flags &= ~(1U << FSP_FLAGS_POS_ENCRYPTION);
+  };
+  return dd_update_tablespace_dd_flags(thd, space_name, update_func);
+}
+
+static bool dd_get_tablespace_flags(THD *thd, const char *space_name,
+                                    uint32_t &dd_space_flags) {
+  const dd::Tablespace *dd_space = nullptr;
+  dd::cache::Dictionary_client::Auto_releaser releaser(thd->dd_client());
+
+  return thd->dd_client()->acquire(space_name, &dd_space) ||
+         dd_space == nullptr ||
+         dd_space->se_private_data().get(dd_space_key_strings[DD_SPACE_FLAGS],
+                                         &dd_space_flags);
+}
+
+bool dd_set_flags(THD *thd, const char *space_name,
+                  const uint32_t space_flags) {
+  auto set_flags = [](uint32_t &dd_space_flags, uint32_t space_flags) {
+    // currently we are using this function only for correcting encryption flag
+    ut_ad(dd_space_flags == space_flags ||
+          FSP_FLAGS_GET_ENCRYPTION(dd_space_flags) !=
+              FSP_FLAGS_GET_ENCRYPTION(space_flags));
+    dd_space_flags = space_flags;
+  };
+  auto update_func = std::bind(set_flags, std::placeholders::_1, space_flags);
+  return dd_update_tablespace_dd_flags(thd, space_name, update_func);
```

Comment:
With the clear and set functions unused, only dd_set_flags is remaining and there is no real reason to use std::function.

Reply:
clear and set functions are used by the encryption threads. I have taken the DD flag update code from there. Turns out that your comment about fixing DD encryption flag for mysql.ibd is only valid for encryption threads. I do not want to put effort in making code changes that will need to be reverted once encryption threads code is merged in. There I will also test the mysql.ibd DD code - so this PR depends on encryption threads anyways.

Reply:
Ok, Do you have testcase for exercising all this code (by encryption threads)?
Should be easy to write one (not as part of this PR )
disable flush, encryption thread encrypt mysql.ibd, crash and ensure startup works fine.

Reply:
yes, this will be added in encryption threads test suite.

Reply:
Does it exist in your encryption branch?

Reply:
one of the tests (more) was failing due to this. I have restarted the server after all tables were encrypted with encryption threads and I got a crash as the DD was not expecting mysql.ibd to be encrypted.